docs/design/coding-conventions.md

```diff
@@ -0,0 +1,413 @@
+
+        case 2:
+        default:
+            x++;
+            mode--;
+            break;
+        }
+    } while(x > 0);
+}
+```
+
+Don't put space between a unary operator and its operand.
+Always put space between an assignment (including compound assignment) operator and its operands.
+For other binary operators, use your best judgement.
+
+If you have to line break a binary expression, put the line break after the operator for an assignment, and before the operator before all others.
+
+Intializer lists for constructors get some special-case formatting rules.
+If everything fits on one line, then great:
+
+```c++
+SomeClass::SomeClass()
+    : a(0), b(1), c(2)
+{}
+```
+
+Otherwise, put the line break *before* the comma:
+
+```c++
+SomeClass::SomeClass()
+    : a(0)
+    , b(1)
+    , c(2)
+{}
+```
+
+When working with `static const` arrays, put the outer `{}` for the array on their own line, and then put each element on its own line and aim for vertical alignment.
+
+```c++
+struct Mapping
+{
+    char const* key;
+    char const* val;
+} kMapping[] =
+{
+    { "a", "aardvark" },
+    { "b", "bat" },
+};
+```
+
+A trailing comma should always be used for array initializer lists like this, as well as for `enum` tags.
+
+When writing multi-line macros, the backslash escapes should align vertically.
+For consistentcy, a trailing `\` can be used, followed by a comment to mark the end of the definition:
+
+```c++
+#define FOREACH_ANIMAL(X)   \
+    X(Aardvark)             \
+    X(Bat)                  \
+    /* end */
+```
+
+Naming
+------
+
+### Casing
+
+Types should in general use `UpperCamelCase`. This includes `struct`s, `class`es, `enum`s and `typedef`s.
+
+Values should in general use `lowerCamelCase`. This includes functions, methods, local variables, global variables, parameters, fields, etc.
+
+Macros should in general use `SCREAMING_SNAKE_CASE`.
+It is important to prefix all macros (e.g., with `SLANG_`) to avoid collisions, since `namespace`s don't affect macros).
+
+### Prefixes
+
+Prefixes based on types (e.g., `p` for pointers) should never be used.
+
+Global variables should have a `g` prefix, e.g. `gCounter`.
+Non-`const` `static` class members can have an `s` prefix if that suits your fancy.
+Of course, both of these should be avoided, so this shouldn't come up often.
+
+Constant data (in the sense of `static const`) should have a `k` prefix.
+
+In `class` types where "information hiding" is relevant/important, an `m` prefix on members is allowed (but not required). Exmaple: `mBegin`.
+In all other cases, member variables/fields should not be prefixed.
+
+No prefixing convention is used to differentiate the public interface of an C++ feature/class from its implementation.
+
+Public C APIs will prefix all symbol names while following the casing convention (e.g. `SlangModule`, `slangLoadModule`, etc.).
+
+### Enums
+
+C-style `enum` should use the following convention:
+
+```c++
+enum Color
+{
+    kColor_Red,
```

Comment:
This seems inconsistent with the enum class case - why have the k prefix on one but not the other. Given the choice not having the prefix on either seems the less noisy, and the _ just replaces :: scoping.